File: src/lib/scala/Slate.Test/src/test/scala/slate/test/QueryTests.scala
/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2016 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */
package slate.test

import org.scalatest.{BeforeAndAfter, BeforeAndAfterAll, FunSpec}
import slate.common.DateTime
import slate.common.query.{QueryEncoder, Query}

class QueryTests extends FunSpec  with BeforeAndAfter with BeforeAndAfterAll {

  before {
    println("before")
  }


  describe ( "Sql Value Conversion") {

    it("can convert null") {
      assert( QueryEncoder.convertVal(null) == "")
    }


    it("can convert string empty") {
      assert( QueryEncoder.convertVal("") == "''")
    }


    it("can convert string non-empty") {
      assert( QueryEncoder.convertVal("slate kit") == "'slate kit'")
    }


    it("can convert string with quote") {
      assert( QueryEncoder.convertVal("slate kit's dope") == "'slate kit''s dope'")
    }


    it("can convert boolean true") {
      assert( QueryEncoder.convertVal(true) == "1")
    }


    it("can convert boolean false") {
      assert( QueryEncoder.convertVal(false) == "0")
    }


    it("can convert boolean datetime") {
      assert( QueryEncoder.convertVal(DateTime(2016, 10, 16)) == "'2016-10-16 12:00:00'")
    }


    it("can convert field") {
      assert( QueryEncoder.ensureField("a(1)2*3&4b") == "a1234b")
    }
  }


  describe ( "Sql Query Builder Filter") {

    it("can build empty") {
      assert( new Query().toFilter() == "")
    }


    it("can build filter 1") {
      assert( new Query().where("name", "=", "slate kit").toFilter() == "name = 'slate kit'")
    }


    it("can build where with 1 field of type bool") {
      assert( new Query().where("isactive", "=", true).toFilter() == "isactive = 1")
    }


    it("can build where with 1 field of type int") {
      assert( new Query().where("status", "=", 3).toFilter() == "status = 3")
    }


    it("can build where with 1 field of type datetime") {
      assert( new Query().where("date", "=", DateTime(2016, 10, 16)).toFilter() == "date = '2016-10-16 12:00:00'")
    }


    it("can convert where with 2 fields") {
      assert( new Query().where("name", "=", "slate kit's").and("version", "=", 2).toFilter()
        == "name = 'slate kit''s' and version = 2")
    }
  }


  describe ( "Sql Update Builder") {

    it("can build empty") {
      assert( new Query().toFilter() == "")
    }


    it("can build where with 1 field of type string") {
      assert( new Query().set("code", "1").where("name", "=", "slate kit").toUpdatesText() == "SET code='1' WHERE name = 'slate kit'")
    }


    it("can build where with 1 field of type bool") {
      assert( new Query().set("code", "1").where("isactive", "=", true).toUpdatesText() == "SET code='1' WHERE isactive = 1")
    }


    it("can build where with 1 field of type int") {
      assert( new Query().set("code", "1").where("status", "=", 3).toUpdatesText() == "SET code='1' WHERE status = 3")
    }


    it("can build where with 1 field of type datetime") {
      assert( new Query().set("code", "1").where("date", "=", DateTime(2016, 10, 16)).toUpdatesText() == "SET code='1' WHERE date = '2016-10-16 12:00:00'")
    }


    it("can convert where with 2 fields") {
      assert( new Query().set("code", "1").where("name", "=", "slate kit's").and("version", "=", 2).toUpdatesText()
        == "SET code='1' WHERE name = 'slate kit''s' and version = 2")
    }
  }
}

File: src/lib/kotlin/slatekit-tests/src/test/kotlin/test/common/QueryTests.kt
/**
<slate_header>
url: www.slatekit.com
git: www.github.com/code-helix/slatekit
org: www.codehelix.co
author: Kishore Reddy
copyright: 2016 CodeHelix Solutions Inc.
license: refer to website and/or github
about: A Kotlin utility library, tool-kit and server backend.
mantra: Simplicity above all else
</slate_header>
 */
package test.common

import org.junit.Assert
import org.junit.Test
import slatekit.common.DateTime
import slatekit.query.Query
import slatekit.query.QueryEncoder
import slatekit.meta.where
import test.setup.Book
import test.setup.StatusEnum


/**
 * Created by kishorereddy on 6/3/17.
 */
class QueryTests {


    @Test fun can_convert_string_empty() {
        Assert.assertTrue( QueryEncoder.convertVal("") == "''")
    }


    @Test fun can_convert_string_non_empty() {
        Assert.assertTrue( QueryEncoder.convertVal("slate kit") == "'slate kit'")
    }


    @Test fun can_convert_string_with_quote() {
        Assert.assertTrue( QueryEncoder.convertVal("slate kit's dope") == "'slate kit''s dope'")
    }


    @Test fun can_convert_boolean_true() {
        Assert.assertTrue( QueryEncoder.convertVal(true) == "1")
    }


    @Test fun can_convert_boolean_false() {
        Assert.assertTrue( QueryEncoder.convertVal(false) == "0")
    }


    @Test fun can_convert_enum() {
        Assert.assertTrue( QueryEncoder.convertVal(StatusEnum.Active) == "1")
    }


    @Test fun can_convert_boolean_datetime() {
        Assert.assertTrue( QueryEncoder.convertVal(DateTime.of(2016, 10, 16)) == "'2016-10-16 00:00:00'")
    }


    @Test fun can_convert_field() {
        Assert.assertTrue( QueryEncoder.ensureField("a(1)2*3&4b") == "a1234b")
    }

    @Test fun can_build_empty() {
        Assert.assertTrue(  Query().toFilter() == "")
    }


    @Test fun can_build_filter_1() {
        Assert.assertTrue(  Query().where("api", "=", "slate kit").toFilter() == "api = 'slate kit'")
    }


    @Test fun can_build_where_with_1_field_of_type_bool() {
        Assert.assertTrue(  Query().where("isactive", "=", true).toFilter() == "isactive = 1")
    }


    @Test fun can_build_where_with_1_field_of_type_int() {
        Assert.assertTrue(  Query().where("status", "=", 3).toFilter() == "status = 3")
    }


    @Test fun can_build_where_with_1_field_of_type_datetime() {
        Assert.assertTrue(  Query().where("date", "=", DateTime.of(2016, 10, 16)).toFilter() == "date = '2016-10-16 00:00:00'")
    }


    @Test fun can_build_where_with_2_fields() {
        Assert.assertTrue(  Query().where("api", "=", "slate kit's").and("version", "=", 2).toFilter()
                == "api = 'slate kit''s' and version = 2")
    }


    @Test fun can_build_where_typed() {
        Assert.assertTrue(  Query().where(Book::rating, ">", 3.0).toFilter()
                == "rating > 3.0")
    }


    @Test fun can_build_condition_is_null() {
        val filter = Query().where(Book::rating, "=", Query.Null).toFilter()
        Assert.assertTrue(  filter == "rating is null")
    }


    @Test fun can_build_condition_is_null_literal() {
        val filter = Query().where(Book::rating, "=", null).toFilter()
        Assert.assertTrue(  filter == "rating is null")
    }


    @Test fun can_build_condition_is_not_null() {
        val filter = Query().where(Book::rating, "!=", Query.Null).toFilter()
        Assert.assertTrue(  filter == "rating is not null")
    }


    @Test fun can_build_condition_in() {
        val filter = Query().where(Book::id, "in", listOf(1,3,5)).toFilter()
        Assert.assertTrue(  filter == "id in (1,3,5)")
    }


    @Test fun can_build_orderby_asc() {
        val filter = Query().orderBy("id", Query.Asc).toFilter()
        Assert.assertTrue(  filter == " order by id asc")
    }


    @Test fun can_build_orderby_desc() {
        val filter = Query().orderBy("id", Query.Desc).toFilter()
        Assert.assertTrue(  filter == " order by id desc")
    }


    @Test fun can_build_join() {
        val filter = Query().join("users", "users.id", "movies.created_by").toFilter()
        Assert.assertTrue(  filter == "  join users on users.id = movies.created_by ")
    }
}